research/gaia/gremlin/gs_gremlin/src/graph_proxy/global_storage.rs

```diff
@@ -0,0 +1,462 @@
+use maxgraph_store::api::graph_schema::Schema;
+use maxgraph_store::api::prelude::Property;
+use maxgraph_store::api::PropId;
+use maxgraph_store::api::*;
+use maxgraph_store::api::{Edge as StoreEdge, Vertex as StoreVertex};
+use std::collections::HashMap;
+use std::sync::Arc;
+
+static INVALID_LABEL_ID: LabelId = 0xffffffff;
+static INVALID_PROP_ID: PropId = 0xffffffff;
+// TODO(bingqing): confirm with compiler, compiler should set the param_key as "SID"
+const SNAPSHOT_ID: &str = "SID";
+
+pub struct GraphScopeStore<V, VI, E, EI>
+where
+    V: StoreVertex + 'static,
+    VI: Iterator<Item = V> + Send + 'static,
+    E: StoreEdge + 'static,
+    EI: Iterator<Item = E> + Send + 'static,
+{
+    store: Arc<dyn GlobalGraphQuery<V = V, E = E, VI = VI, EI = EI>>,
+    partition_manager: Arc<dyn GraphPartitionManager>,
+}
+
+#[allow(dead_code)]
+pub fn create_gs_store<V, VI, E, EI>(
+    store: Arc<dyn GlobalGraphQuery<V = V, E = E, VI = VI, EI = EI>>,
+    partition_manager: Arc<dyn GraphPartitionManager>,
+) where
+    V: StoreVertex + 'static,
+    VI: Iterator<Item = V> + Send + 'static,
+    E: StoreEdge + 'static,
+    EI: Iterator<Item = E> + Send + 'static,
+{
+    let graph = GraphScopeStore {
+        store,
+        partition_manager,
+    };
+    register_graph(Arc::new(graph));
+}
+
+impl<V, VI, E, EI> GraphProxy for GraphScopeStore<V, VI, E, EI>
+where
+    V: StoreVertex + 'static,
+    VI: Iterator<Item = V> + Send + 'static,
+    E: StoreEdge + 'static,
+    EI: Iterator<Item = E> + Send + 'static,
+{
+    fn scan_vertex(
+        &self,
+        params: &QueryParams<Vertex>,
+    ) -> DynResult<Box<dyn Iterator<Item = Vertex> + Send>> {
+        let store = self.store.clone();
+        let si = params
+            .get_extra_param(SNAPSHOT_ID)
+            .ok_or(str_to_dyn_error("get snapshot_id failed"))?
+            .as_i64()
+            .map_err(|e| str_to_dyn_error(&e.to_string()))? as SnapshotId;
+        let schema = store
+            .get_schema(si)
+            .ok_or(str_to_dyn_error("get schema failed"))?;
+        let label_ids = encode_storage_label(params.labels.as_ref(), schema.clone());
+        let prop_ids = encode_storage_prop_key(params.props.as_ref(), schema.clone());
+        let filter = params.filter.clone();
+        let result = store
+            .get_all_vertices(
+                si,
+                label_ids.as_ref(),
+                // None means no filter condition pushed down to storage as not supported yet. Same as follows.
+                None,
+                // None means no need to dedup by properties. Same as follows.
+                None,
+                prop_ids.as_ref(),
+                // Zero limit means no limit. Same as follows.
+                params.limit.unwrap_or(0),
+                // TODO(bingqing): optimization, each worker may scan some partitions on current server.
+                // Empty vector means all partitions, since we use worker 0 to scan all for now.
```

Comment:
The comment looks strange to me. worker 0 for now? Why? What is your future plan?

Reply:
We assign workers to scan partitions in source operator, and for now, worker 0 (specifically,  the first worker on each server) is going to scan all partitions on current server for correctness guarantee. In the future, source op should assign each worker to scan some partitions, i.e., allow parallel scan.

Reply:
BTW, I didn't do this optimization since V2 store and vineyard store seems to give the info of "which worker should scan which partitions" in a ununified way. I will do this optimization after adapting to vineyard.